GUI Refactoring : Share elements of link upper banner between pages

guilpier-code opened the pull request:
See message of commit  "Refactoring link upper banner : first try ==> failure at this stage" (30cbcc082a5107ed10a88fa71d0351b43f9eac23)

guilpier-code committed:
Link upper banner : update the banner (or board) when changing of tab

When we change a link property (transmission capacity usage, hurdle costs usage, ...) in the upper banner, and than
change of tab  (for example from Transmission capacities to Parameters), his upper banner is not properly refreshed.
This is fixed here.

flomnes committed:
Add const here and there

guilpier-code committed:
Update link upper banner (or board) when changing of tab : fix a remaining bug

guilpier-code committed:
Refactoring link upper banner : first try ==> failure at this stage

We tried to make the transmission capacities usage button and its behavior a unique object shared by
the 2 pages "Parameters" and "Transmission capacities".
This try compiles but fails at execution : when changing the value of the capacities usage on one page,
we cannot refresh this change on the other page. So we get back to where we used to be.

flomnes commented:
I tried the same modification as your 1st try, and it also failed. Let's try to merge this in 8.2.1

guilpier-code committed:
Refactoring link upper banner : make NTC usage button a class

guilpier-code committed:
Refactoring link upper banner : put NTC usage button class in a separate file

guilpier-code committed:
Refactoring link upper banner : cleaning and renaming

guilpier-code committed:
Refactoring link upper banner : label "Link" is separate from caption button.

guilpier-code committed:
Refactoring link upper banner : create a caption button class and put it in the "button" file

guilpier-code committed:
Refactoring link upper banner : create a loop flow button class and put it in the "button" file

guilpier-code committed:
Refactoring link upper banner : a bit of cleaning

guilpier-code committed:
Refactoring link upper banner : creation of button parent class to factorize some code

guilpier-code committed:
Refactoring link upper banner : button parent class renamed

guilpier-code committed:
Refactoring link upper banner : factorizing broadcast and destruction

guilpier-code committed:
Refactoring link upper banner : extracting a phase button class

guilpier-code committed:
[skip ci] Refactoring link upper banner : extracting classes for asset type and hurdle costs usage

guilpier-code committed:
Refactoring link upper banner : a bit of cleaning

guilpier-code committed:
Squashed commit of the following:

commit f832da3b7f132e44b83f5b723ad9af3c88458232
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Tue Mar 1 16:36:16 2022 +0100

    Pressing the 'delete' opens a deletion popup for current selection (#562)

    * Pressing the 'delete' opens a deletion popup for current selection

    * Remove unused header inclusion

    * Reset new pointer in Control::reset()

    * Templatize Component::evtNewArea

    * Set focus & reset pointer

    * Reset pointer

commit f7a3fc06dee11b4610634ba17a73291a78686519
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Tue Mar 1 16:16:12 2022 +0100

    Set RC=0

commit 8f8b042478890a9f7fb0c7c1b403534708bca069
Merge: f3f20565 1a759bff
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Tue Mar 1 16:15:35 2022 +0100

    Merge pull request #583 from AntaresSimulatorTeam/release/8.2

    Release/8.2

commit f3f205651d4f659352add08ac5d07e70bea601ad
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:33:34 2022 +0100

    [vcpkg] Update submodule 2021.05.12 -> 2022-02-01

commit fb810e1060dbfe38141031c27dac60fd827a0558
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:14:08 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2 (ter)

commit edb8ff928692e0a2777c7e56d20c968d2e7d5a3e
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:09:09 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2 (bis)

commit 306e59f5fc7c18fae0dcd4a6ab9d05cccec02b5a
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:09:09 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2

commit 57012d995519021e28aa7a9f534347f559868a46
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:00:28 2022 +0100

    Fix bug related to short-circuit evaluation (#576)

    * Fix bug related to short-circuit evaluation

    * Fix short-circuit evaulation in TSnumberData.cpp

commit 3ae40e596897e1307ce09b2e9a9ac73279ef5162
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 10:10:47 2022 +0100

    [skip ci] Set tag RC2

commit 17e3a0f4cb620890b9323e050a5145269c8ec7d6
Author: Florian Omnès <florian@omnes.cc>
Date:   Fri Feb 25 09:32:46 2022 +0100

    Change styles related to loop-flow (#571)

    * Change styles related to loop-flow

    * Change red (error) to orange (warning)
    * Warning only if loop flow is enabled
    * Cell style "warning" for transmission capacites if rule is violated

    * PR remarks

    * PR remarks

    * Turn static checkLoopFlow functions into virtual methods

commit 8969158f8d0f29588ed7a91e9775d118ef9c4505
Author: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Date:   Thu Feb 24 16:32:46 2022 +0100

    Link upper banner : update the banner (or board) when changing of tab (#572)

    * Link upper banner : update the banner (or board) when changing of tab

    When we change a link property (transmission capacity usage, hurdle costs usage, ...) in the upper banner, and than
    change of tab  (for example from Transmission capacities to Parameters), his upper banner is not properly refreshed.
    This is fixed here.

    * Add const here and there

    * Update link upper banner (or board) when changing of tab : fix a remaining bug

    Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

commit 335157136cf611bd7d0326f79a4245171e191090
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:56 2022 +0100

    Fix infinite capacity bug (#568)

    * Fix infinite capacity bug

    * Remove redundant "+" signs

commit 1231b111765a8b33ca1482eac98ea42fb4241fe9
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:33 2022 +0100

    Remove optImmediate from links' timeseries (#552)

commit 6e62414571ed7ef2adfbd10b514a18147f05d610
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:25 2022 +0100

    Ignore disabled MC years when computing average NTC (#558)

commit 89349618886e2608d554f40522ef49030c62e2a1
Author: Florian Omnès <florian@omnes.cc>
Date:   Mon Feb 21 21:10:26 2022 +0100

    Revert "UI: Change and rename "upstream/downstream" view (#468)" (#570)

    This reverts commit dec594b6ee685e1f7ed1368fb856a8f35b90126b.

commit 9874d49862eb8e284b8a42cd5383af8dd21d695b
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 16 17:29:37 2022 +0100

    [UI] Fix segfault on area copy-paste (#550)

commit ce53dfcf4369c04de3d88e770d14a59ea9f92a8e
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Thu Dec 16 11:26:29 2021 +0100

    [CI] Fix CentOS release

commit cfab3c4bdd6d6269c8764b7538f27b06e69bcf69
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 17:58:55 2022 +0100

    [skip ci] Remove Changelog related to hydro change

commit 1219fbed1c98b94982b09e4427f6c9e22a4175ab
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 17:58:55 2022 +0100

    [skip ci] Remove Changelog related to hydro change

commit a70f72a49701f11c7d433955592c4c8152a28557
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 16:56:20 2022 +0100

    Skip failed nonreg tests for Linux (bis)

commit ec090c694f7c72468643623ae72f6a7dd8bef703
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 16:27:52 2022 +0100

    Skip failed nonreg tests for Linux

commit 48a813061585b876e8a437599cc143b3f22bd046
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 16:11:30 2022 +0100

    Skip failed tests on Windows

commit d41b1c40f47e28314df269acc6d862af076b3cd9
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 16:09:16 2022 +0100

    Revert "Include water level equations whenever hydro-pricing==accurate or heuristic==false or pumping==true (#524)"

    This reverts commit cfd9100a33b60a738b217db2dda4839bb9ddcc9f.

commit ed938bdd2de857104e322e0f4d8e698fe8ee9ee4
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 16:08:07 2022 +0100

    Revert "Ignore failed tests on Ubuntu/CentOS"

    This reverts commit 60d07f90dbee36f6b197e20b6234dfb59b793120.

commit ffb46a41632e330e37411898952f42751fd57fe1
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Wed Feb 9 13:46:22 2022 +0100

    Ignore failed tests on Ubuntu/CentOS

    Add TODO note

commit b4b7a4c0b3bbc6d5b1b3ab14e59b19a7609677de
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Tue Feb 8 14:28:57 2022 +0100

    Add changelog for 8.2.0-rc1

commit 328d2227ab67bd85506e69d99375e379a5c42fda
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Tue Feb 8 13:52:16 2022 +0100

    Add changelog for v8.1.1

    Had been forgotten, oops !

commit 1a759bff9a234f87c294cd679105e69713105cd4
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:33:34 2022 +0100

    [vcpkg] Update submodule 2021.05.12 -> 2022-02-01

commit f1431784ee43943001ef7ca3f590431cf1b5fa97
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:14:08 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2 (ter)

commit d2b1a2e6f0ca883de827542987b0be98202ded01
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:09:09 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2 (bis)

commit 15d1dd3d861523fac04407502257241b76dc6831
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:09:09 2022 +0100

    [skip ci] Update changelog for v8.2.0-rc2

commit 5b2a3e0d07e75a50c9af6b04009e1ee21f21c57e
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Fri Feb 25 14:00:28 2022 +0100

    Fix bug related to short-circuit evaluation (#576)

    * Fix bug related to short-circuit evaluation

    * Fix short-circuit evaulation in TSnumberData.cpp

commit 7be9347fbd093cd95b810e7f00b5e656956ff4e1
Author: Florian OMNES <florian.omnes@rte-france.com>
Date:   Fri Feb 25 10:10:47 2022 +0100

    [skip ci] Set tag RC2

commit d608e5635f132d4c0d35314e9a54cc2472787539
Author: Florian Omnès <florian@omnes.cc>
Date:   Fri Feb 25 09:32:46 2022 +0100

    Change styles related to loop-flow (#571)

    * Change styles related to loop-flow

    * Change red (error) to orange (warning)
    * Warning only if loop flow is enabled
    * Cell style "warning" for transmission capacites if rule is violated

    * PR remarks

    * PR remarks

    * Turn static checkLoopFlow functions into virtual methods

commit 402c5bbef62c670918f27c877c9916159323e1f9
Author: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Date:   Thu Feb 24 16:32:46 2022 +0100

    Link upper banner : update the banner (or board) when changing of tab (#572)

    * Link upper banner : update the banner (or board) when changing of tab

    When we change a link property (transmission capacity usage, hurdle costs usage, ...) in the upper banner, and than
    change of tab  (for example from Transmission capacities to Parameters), his upper banner is not properly refreshed.
    This is fixed here.

    * Add const here and there

    * Update link upper banner (or board) when changing of tab : fix a remaining bug

    Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

commit 79acbbd32767335fe55f76c7a82c16c67a2fb59c
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:56 2022 +0100

    Fix infinite capacity bug (#568)

    * Fix infinite capacity bug

    * Remove redundant "+" signs

commit 0997ca775acd719f425f7ad15daf20fcf08ea029
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:33 2022 +0100

    Remove optImmediate from links' timeseries (#552)

commit e884766566c4c67c50228665cb27f1ffbb8bb377
Author: Florian Omnès <florian.omnes@rte-france.com>
Date:   Wed Feb 23 14:03:25 2022 +0100

    Ignore disabled MC years when computing average NTC (#558)

# Conflicts:
#	src/ui/simulator/windows/connection.cpp
#	src/ui/simulator/windows/connection.h

guilpier-code committed:
Refactoring link upper banner : small correction (cancel a trace of a previous fruitless development)

guilpier-code committed:
Refactoring link upper banner : corrections due to sonarQ review

guilpier-code committed:
Refactoring link upper banner : bind button to menu + create button method

- Factor code that binds button to a menu (see bindButtonToPopupMenu() function)
- Isolate a button creation in a polymorphic function named makeButton(...)

guilpier-code committed:
Refactoring link upper banner : cleaning due to sonarQ review

guilpier-code committed:
Refactoring link upper banner : corrections due to sonarQ review

guilpier-code committed:
Refactoring link upper banner : small corrections to lower number of sonarQ remarks

flomnes reviewed (commented):
Maybe instantiate the Yuni::Event in file create.cpp ?
```cpp
// Event observer instantiation
Yuni::Event<...> observer;
// links parameters time series
auto* parametersGrid = new_check_allocation<Window::linkParametersGrid>();
auto* intercoParam = new_check_allocation<Window::Interconnection>(page.first, page.second, parametersGrid, observer);
pageLinksParameters = page.first->add(intercoParam, wxT(" Parameters "));
 // links NTC time series
auto* ntcGrid = new_check_allocation<Window::linkNTCgrid>();
auto* intercoGrid = new_check_allocation<Window::Interconnection>(page.first, page.second, ntcGrid, observer);
```

I have no other remark.

guilpier-code committed:
Refactoring link upper banner : remove useless things

flomnes committed:
Fix minor code smells
